File: backend/src/models/prestasi.go
package models

import (
	"time"

	"gorm.io/gorm"
)

// Prestasi merepresentasikan data capaian juara dan penghargaan siswa SKOMDA.
type Prestasi struct {
	ID           uint           `gorm:"primaryKey" json:"id"`
	Slug         string         `gorm:"size:255;not null;uniqueIndex" json:"slug"`
	Title        string         `gorm:"size:255;not null" json:"title"`
	Category     string         `gorm:"size:50;not null;index" json:"category"` // IT & AI, Olahraga, Seni & Kreatif, Kepemimpinan
	Award        string         `gorm:"size:100;not null" json:"award"`
	BadgeLevel   string         `gorm:"size:50;not null" json:"badgeLevel"` // Juara 1, Juara 2, Juara 3, Gold Medal
	Competition  string         `gorm:"size:255;not null" json:"competition"`
	Organizer    string         `gorm:"size:255;not null" json:"organizer"`
	Year         string         `gorm:"size:10;not null;index" json:"year"`
	StudentName  string         `gorm:"size:150;not null" json:"studentName"`
	StudentClass string         `gorm:"size:50;not null" json:"studentClass"`
	Image        string         `gorm:"size:500" json:"image"`
	Description  string         `gorm:"type:text" json:"description"`
	CreatedAt    time.Time      `json:"created_at"`
	UpdatedAt    time.Time      `json:"updated_at"`
	DeletedAt    gorm.DeletedAt `gorm:"index" json:"-"`
}

File: frontend/src/services/prestasi.ts
export interface PrestasiItem {
  id?: number | string;
  slug?: string;
  title: string;
  category: "IT & AI" | "Olahraga" | "Seni & Kreatif" | "Kepemimpinan" | string;
  award: string;
  badgeLevel: string;
  competition: string;
  organizer: string;
  year: string;
  studentName: string;
  studentClass: string;
  image?: string;
  description?: string;
  created_at?: string;
}

export const PRESTASI_CATEGORIES = [
  "Semua",
  "IT & AI",
  "Olahraga",
  "Seni & Kreatif",
  "Kepemimpinan",
] as const;

const API_BASE_URL =
  process.env.NEXT_PUBLIC_API_URL || "http://localhost:8080/api";

export async function getPrestasiList(
  category?: string,
  year?: string
): Promise<PrestasiItem[]> {
  try {
    const url = new URL(`${API_BASE_URL}/prestasi`);
    if (category && category !== "Semua") {
      url.searchParams.set("category", category);
    }
    if (year && year !== "Semua") {
      url.searchParams.set("year", year);
    }
    const res = await fetch(url.toString(), { cache: "no-store" });
    if (res.ok) {
      const json = await res.json();
      return json.data || [];
    }
    return [];
  } catch {
    return [];
  }
}

export async function createPrestasi(
  data: Partial<PrestasiItem>
): Promise<{ success: boolean; data?: PrestasiItem; error?: string }> {
  try {
    const res = await fetch(`${API_BASE_URL}/prestasi`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      credentials: "include",
      body: JSON.stringify(data),
    });
    const json = await res.json();
    if (!res.ok) return { success: false, error: json.error || "Gagal menyimpan prestasi" };
    return { success: true, data: json.data };
  } catch (err: any) {
    return { success: false, error: err.message || "Gagal terhubung ke backend" };
  }
}

export async function updatePrestasi(
  id: number | string,
  data: Partial<PrestasiItem>
): Promise<{ success: boolean; data?: PrestasiItem; error?: string }> {
  try {
    const res = await fetch(`${API_BASE_URL}/prestasi/${id}`, {
      method: "PUT",
      headers: { "Content-Type": "application/json" },
      credentials: "include",
      body: JSON.stringify(data),
    });
    const json = await res.json();
    if (!res.ok) return { success: false, error: json.error || "Gagal memperbarui prestasi" };
    return { success: true, data: json.data };
  } catch (err: any) {
    return { success: false, error: err.message || "Gagal terhubung ke backend" };
  }
}

export async function deletePrestasi(
  id: number | string
): Promise<{ success: boolean; error?: string }> {
  try {
    const res = await fetch(`${API_BASE_URL}/prestasi/${id}`, {
      method: "DELETE",
      credentials: "include",
    });
    const json = await res.json();
    if (!res.ok) return { success: false, error: json.error || "Gagal menghapus prestasi" };
    return { success: true };
  } catch (err: any) {
    return { success: false, error: err.message || "Gagal terhubung ke backend" };
  }
}
